share/man/man4/video.4
.\" $OpenBSD: video.4,v 1.20 2022/03/31 17:27:21 naddy Exp $
.\"
.\" Copyright (c) 2008 Marcus Glocker <mglocker@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: March 31 2022 $
.Dt VIDEO 4
.Os
.Sh NAME
.Nm video
.Nd device-independent video driver layer
.Sh SYNOPSIS
.Cd "video* at uvideo?"
.Pp
.In sys/types.h
.In sys/ioctl.h
.In sys/videoio.h
.Sh DESCRIPTION
The
.Nm
driver provides support for various video devices.
It provides a uniform programming interface layer
above different underlying video hardware drivers.
The
.Nm
driver uses the V4L2 (Video for Linux Two) API which is widely used by video
applications.
Therefore this document mainly describes the V4L2 API parts
which are supported by the
.Nm
driver.
.Pp
For security reasons video recording is blanked by default.
To achieve this, the
.Nm
driver blanks image data received from the underlying video hardware driver.
The superuser can change this behavior using the
.Va kern.video.record
.Xr sysctl 2
variable:
.Bd -literal -offset indent
kern.video.record=0	# Recording is blanked (default)
kern.video.record=1	# Recording is enabled
.Ed
.Sh IOCTLS
The following
.Xr ioctl 2
commands are supported:
.Bl -tag -width Ds
.It Dv VIDIOC_QUERYCAP Fa "struct v4l2_capability *"
Query device capabilities.
.Bd -literal
struct v4l2_capability {
	u_int8_t	driver[16];
	u_int8_t	card[32];
	u_int8_t	bus_info[32];
	u_int32_t	version;
	u_int32_t	capabilities;
	u_int32_t	device_caps;
	u_int32_t	reserved[3];
};
.Ed
.It Dv VIDIOC_ENUM_FMT Fa "struct v4l2_fmtdesc *"
Enumerate image formats.
.Bd -literal
struct v4l2_fmtdesc {
	u_int32_t	index;
	u_int32_t	type;
	u_int32_t	flags;
	u_int8_t	description[32];
	u_int32_t	pixelformat;
	u_int32_t	mbus_code;
	u_int32_t	reserved[3];
};
.Ed
.It Dv VIDIOC_S_FMT Fa "struct v4l2_format *"
Set the data format.
.Bd -literal
struct v4l2_format {
	u_int32_t	type;
	union {
		struct v4l2_pix_format		pix;
		struct v4l2_pix_format_mplane	pix_mp;
		struct v4l2_window		win;
		struct v4l2_vbi_format		vbi;
		struct v4l2_sliced_vbi_format	sliced;
		struct v4l2_sdr_format		sdr;
		struct v4l2_meta_format		meta;
		u_int8_t			raw_data[200];
        } fmt;
};
.Ed
.It Dv VIDIOC_G_FMT Fa "struct v4l2_format *"
Get the data format.
.Pp
Same structure as for
.Dv VIDIOC_S_FMT .
.It Dv VIDIOC_ENUMINPUT Fa "struct v4l2_input *"
Enumerate video inputs.
.Bd -literal
struct v4l2_input {
	u_int32_t	index;
	u_int8_t	name[32];
	u_int32_t	type;
	u_int32_t	audioset;
	u_int32_t	tuner;
	v4l2_std_id	std;
	u_int32_t	status;
	u_int32_t	capabilities;
	u_int32_t	reserved[3];
};
.Ed
.It Dv VIDIOC_G_INPUT Fa "int *"
Get the current video input.
.It Dv VIDIOC_S_INPUT Fa "int *"
Select the current video input.
.It Dv VIDIOC_REQBUFS Fa "struct v4l2_requestbuffers *"
Initiate memory mapping or user pointer I/O.
.Bd -literal
struct v4l2_requestbuffers {
	u_int32_t	count;
	u_int32_t	type;
	u_int32_t	memory;
	u_int32_t	capabilities;
	u_Int32_t	flags;
	u_int32_t	reserved[3];
};
.Ed
.It Dv VIDIOC_QUERYBUF Fa "struct v4l2_buffer *"
Query the status of a buffer.
.Bd -literal
struct v4l2_buffer {
	u_int32_t		index;
	u_int32_t		type;
	u_int32_t		bytesused;
	u_int32_t		flags;
	u_int32_t		field;
	struct timeval		timestamp;
	struct v4l2_timecode	timecode;
	u_int32_t		sequence;
	u_int32_t		memory;
	union {
		u_int32_t		offset;
		unsigned long		userptr;
		struct v4l2_plane	*planes;
		int32_t			fd;
	} m;
	u_int32_t		length;
	u_int32_t		reserved2;
	union {
		int32_t		request_fd;
		u_int32_t	reserved;
	}
};
.Ed
.It Dv VIDIOC_QBUF Fa "struct v4l2_buffer *"
Add a buffer to the queue.
.Pp
Same structure as for
.Dv VIDIOC_QUERYBUF .
.It Dv VIDIOC_DQBUF Fa "struct v4l2_buffer *"
Remove a buffer from the queue.
.Pp
Same structure as for
.Dv VIDIOC_QUERYBUF .
.It Dv VIDIOC_STREAMON Fa "int *"
Start video stream.
.It Dv VIDIOC_STREAMOFF Fa "int *"
Stop video stream.
.It Dv VIDIOC_TRY_FMT Fa "struct v4l2_format *"
Try a data format.
.Pp
Same structure as for
.Dv VIDIOC_S_FMT .
.It Dv VIDIOC_ENUM_FRAMEINTERVALS Fa "struct v4l2_frmivalenum *"
Enumerate frame intervals.
.Bd -literal
struct v4l2_frmivalenum {
	u_int32_t		index;
	u_int32_t		pixel_format;
	u_int32_t		width;
	u_int32_t		height;
	u_int32_t		type;
	union {
		struct v4l2_fract		discrete;
		struct v4l2_frmival_stepwise	stepwise;
	};
	u_int32_t		reserved[2];
};

struct v4l2_frmival_stepwise {
	struct v4l2_fract min;
	struct v4l2_fract max;
	struct v4l2_fract step;
};
.Ed
.It Dv VIDIOC_S_PARM Fa "struct v4l2_streamparm *"
Set streaming parameters.
.Bd -literal
struct v4l2_streamparm {
	u_int32_t	type;
	union {
		struct v4l2_captureparm	capture;
		struct v4l2_outputparm	output;
		u_int8_t		raw_data[200];
	} parm;
};

struct v4l2_captureparm	{
	u_int32_t	capability;
	u_int32_t	capturemode;
	struct v4l2_fract	timeperframe;
	u_int32_t	extendedmode;
	u_int32_t	readbuffers;
	u_int32_t	reserved[4];
};

struct v4l2_outputparm	{
	u_int32_t	capability;
	u_int32_t	outputmode;
	struct v4l2_fract	timeperframe;
	u_int32_t	extendedmode;
	u_int32_t	writebuffers;
	u_int32_t	reserved[4];
};
.Ed
.It Dv VIDIOC_G_PARM Fa "struct v4l2_streamparm *"
Get the streaming parameters.
.Pp
Same structures as for
.Dv VIDIOC_S_PARM .
.It Dv VIDIOC_QUERYCTRL Fa "struct v4l2_queryctrl *"
Enumerate control items.
.Bd -literal
struct v4l2_queryctrl {
	u_int32_t	id;
	u_int32_t	type;
	u_int8_t	name[32];
	int32_t		minimum;
	int32_t		maximum;
	int32_t		step;
	int32_t		default_value;
	u_int32_t	flags;
	u_int32_t	reserved[2];
};
.Ed
.El
.Pp
Command independent enumerations are:
.Bd -literal
enum v4l2_buf_type {
	V4L2_BUF_TYPE_VIDEO_CAPTURE = 1,
	V4L2_BUF_TYPE_VIDEO_OUTPUT = 2,
	V4L2_BUF_TYPE_VIDEO_OVERLAY = 3,
	V4L2_BUF_TYPE_VBI_CAPTURE = 4,
	V4L2_BUF_TYPE_VBI_OUTPUT = 5,
	V4L2_BUF_TYPE_SLICED_VBI_CAPTURE = 6,
	V4L2_BUF_TYPE_SLICED_VBI_OUTPUT = 7,
	V4L2_BUF_TYPE_VIDEO_OUTPUT_OVERLAY = 8,
	V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE = 9,
	V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE = 10,
	V4L2_BUF_TYPE_SDR_CAPTURE = 11,
	V4L2_BUF_TYPE_SDR_OUTPUT = 12,
	V4L2_BUF_TYPE_META_CAPTURE = 13,
	V4L2_BUF_TYPE_META_OUTPUT = 14,
	/* Deprecated, do not use */
	V4L2_BUF_TYPE_PRIVATE = 0x80,
};

enum v4l2_memory {
	V4L2_MEMORY_MMAP = 1,
	V4L2_MEMORY_USERPTR = 2,
	V4L2_MEMORY_OVERLAY = 3,
	V4L2_MEMORY_DMABUF = 4,
};

enum v4l2_ctrl_type {
	V4L2_CTRL_TYPE_INTEGER = 1,
	V4L2_CTRL_TYPE_BOOLEAN = 2,
	V4L2_CTRL_TYPE_MENU = 3,
	V4L2_CTRL_TYPE_BUTTON = 4,
	V4L2_CTRL_TYPE_INTEGER64 = 5,
	V4L2_CTRL_TYPE_CTRL_CLASS = 6,
	V4L2_CTRL_TYPE_STRING = 7,
	V4L2_CTRL_TYPE_BITMASK = 8,
	V4L2_CTRL_TYPE_INTEGER_MENU = 9,
	V4L2_CTRL_COMPOUND_TYPES = 0x0100,
	V4L2_CTRL_TYPE_U8 = 0x0100,
	V4L2_CTRL_TYPE_U16 = 0x0101,
	V4L2_CTRL_TYPE_U32 = 0x0102,
	V4L2_CTRL_TYPE_AREA = 0x0106,

	/* Compound types are >= 0x0100 */
	V4L2_CTRL_TYPE_HDR10_CLL_INFO = 0x0110,
	V4L2_CTRL_TYPE_HDR10_MASTERING_DISPLAY = 0x0111,

	V4L2_CTRL_TYPE_H264_SPS = 0x0200,
	V4L2_CTRL_TYPE_H264_PPS = 0x0201,
	V4L2_CTRL_TYPE_H264_SCALING_MATRIX = 0x0202,
	V4L2_CTRL_TYPE_H264_SLICE_PARAMS = 0x0203,
	V4L2_CTRL_TYPE_H264_DECODE_PARAMS = 0x0204,
	V4L2_CTRL_TYPE_H264_PRED_WEIGHTS = 0x0205,

	V4L2_CTRL_TYPE_FWHT_PARAMS = 0x0220,

	V4L2_CTRL_TYPE_VP8_FRAME = 0x0240,

	V4L2_CTRL_TYPE_MPEG2_QUANTISATION = 0x0250,
	V4L2_CTRL_TYPE_MPEG2_SEQUENCE = 0x0251,
	V4L2_CTRL_TYPE_MPEG2_PICTURE = 0x0252,

	V4L2_CTRL_TYPE_VP9_COMPRESSED_HDR = 0x0260,
	V4L2_CTRL_TYPE_VP9_FRAME = 0x0261,

	V4L2_CTRL_TYPE_HEVC_SPS = 0x0270,
	V4L2_CTRL_TYPE_HEVC_PPS = 0x0271,
	V4L2_CTRL_TYPE_HEVC_SLICE_PARAMS = 0x0272,
	V4L2_CTRL_TYPE_HEVC_SCALING_MATRIX = 0x0273,
	V4L2_CTRL_TYPE_HEVC_DECODE_PARAMS = 0x0274,

	V4L2_CTRL_TYPE_AV1_SEQUENCE = 0x280,
	V4L2_CTRL_TYPE_AV1_TILE_GROUP_ENTRY = 0x281,
	V4L2_CTRL_TYPE_AV1_FRAME = 0x282,
	V4L2_CTRL_TYPE_AV1_FILM_GRAIN = 0x283,
};

enum v4l2_frmivaltypes {
	V4L2_FRMIVAL_TYPE_DISCRETE = 1,
	V4L2_FRMIVAL_TYPE_CONTINUOUS = 2,
	V4L2_FRMIVAL_TYPE_STEPWISE = 3,
};
.Ed
.Pp
Command independent structures are:
.Bd -literal
struct v4l2_pix_format {
	u_int32_t	width;
	u_int32_t	height;
	u_int32_t	pixelformat;
	u_int32_t	field;
	u_int32_t	bytesperline;
	u_int32_t	sizeimage;
	u_int32_t	colorspace;
	u_int32_t	priv;
	u_int32_t	flags;
	union {
		u_int32_t	ycbcr_enc;
		u_int32_t	hsv_enc;
	};
	u_int32_t	quantization;
	u_int32_t	xfer_func;
};

struct v4l2_window {
	struct v4l2_rect	w;
	u_int32_t		field;
	u_int32_t		chromakey;
	struct v4l2_clip	*clips;
	u_int32_t		clipcount;
	void			__user *bitmap;
	u_int8_t		global_alpha;
};

struct v4l2_vbi_format {
	u_int32_t		sampling_rate;
	u_int32_t		offset;
	u_int32_t		samples_per_line;
	u_int32_t		sample_format;
	int32_t			start[2];
	u_int32_t		count[2];
	u_int32_t		flags;
	u_int32_t		reserved[2];
};

struct v4l2_sliced_vbi_format {
	u_int16_t	service_set;
	u_int16_t	service_lines[2][24];
	u_int32_t	io_size;
	u_int32_t	reserved[2];
};

struct v4l2_fract {
	u_int32_t	numerator;
	u_int32_t	denominator;
};
.Ed
.Pp
Command independent typedefs are:
.Bd -literal
typedef u_int64_t	v4l2_std_id;
.Ed
.Sh READ
Video data can be accessed via the
.Xr read 2
system call.

sys/dev/video.c
/*	$OpenBSD: video.c,v 1.59 2024/12/16 21:22:51 mvs Exp $	*/

/*
 * Copyright (c) 2008 Robert Nagy <robert@openbsd.org>
 * Copyright (c) 2008 Marcus Glocker <mglocker@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/errno.h>
#include <sys/event.h>
#include <sys/ioctl.h>
#include <sys/fcntl.h>
#include <sys/device.h>
#include <sys/vnode.h>
#include <sys/kernel.h>
#include <sys/malloc.h>
#include <sys/mutex.h>
#include <sys/conf.h>
#include <sys/proc.h>
#include <sys/videoio.h>

#include <dev/video_if.h>

#include <uvm/uvm_extern.h>

/*
 * Locks used to protect struct members and global data
 *	a	atomic
 *	m	sc_mtx
 */

#ifdef VIDEO_DEBUG
int video_debug = 1;
#define DPRINTF(l, x...) do { if ((l) <= video_debug) printf(x); } while (0)
#else
#define DPRINTF(l, x...)
#endif

struct video_softc {
	struct device		 dev;
	void			*hw_hdl;	/* hardware driver handle */
	struct device		*sc_dev;	/* hardware device struct */
	const struct video_hw_if *hw_if;	/* hardware interface */
	char			 sc_dying;	/* device detached */
	struct process		*sc_owner;	/* owner process */
	uint8_t			 sc_open;	/* device opened */

	struct mutex		 sc_mtx;
	int			 sc_fsize;
	uint8_t			*sc_fbuffer;
	caddr_t			 sc_fbuffer_mmap;
	size_t			 sc_fbufferlen;
	int			 sc_vidmode;	/* access mode */
#define		VIDMODE_NONE	0
#define		VIDMODE_MMAP	1
#define		VIDMODE_READ	2
	int			 sc_frames_ready;	/* [m] */

	struct klist		 sc_rklist;		/* [m] read selector */
};

int	videoprobe(struct device *, void *, void *);
void	videoattach(struct device *, struct device *, void *);
int	videodetach(struct device *, int);
int	videoactivate(struct device *, int);
int	videoprint(void *, const char *);

void	video_intr(void *);
int	video_stop(struct video_softc *);
int	video_claim(struct video_softc *, struct process *);

const struct cfattach video_ca = {
	sizeof(struct video_softc), videoprobe, videoattach,
	videodetach, videoactivate
};

struct cfdriver video_cd = {
	NULL, "video", DV_DULL
};

/*
 * Global flag to control if video recording is enabled by kern.video.record.
 */
int video_record_enable = 0;	/* [a] */

int
videoprobe(struct device *parent, void *match, void *aux)
{
	return (1);
}

void
videoattach(struct device *parent, struct device *self, void *aux)
{
	struct video_softc *sc = (void *)self;
	struct video_attach_args *sa = aux;

	printf("\n");
	sc->hw_if = sa->hwif;
	sc->hw_hdl = sa->hdl;
	sc->sc_dev = parent;
	sc->sc_fbufferlen = 0;
	sc->sc_owner = NULL;
	mtx_init(&sc->sc_mtx, IPL_MPFLOOR);
	klist_init_mutex(&sc->sc_rklist, &sc->sc_mtx);

	if (sc->hw_if->get_bufsize)
		sc->sc_fbufferlen = (sc->hw_if->get_bufsize)(sc->hw_hdl);
	if (sc->sc_fbufferlen == 0) {
		printf("video: could not request frame buffer size\n");
		return;
	}

	sc->sc_fbuffer = malloc(sc->sc_fbufferlen, M_DEVBUF, M_NOWAIT);
	if (sc->sc_fbuffer == NULL) {
		printf("video: could not allocate frame buffer\n");
		return;
	}
}

int
videoopen(dev_t dev, int flags, int fmt, struct proc *p)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	int error = 0;

	KERNEL_ASSERT_LOCKED();

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (ENXIO);

	if (sc->sc_open) {
		DPRINTF(1, "%s: device already open\n", __func__);
		goto done;
	}

	sc->sc_vidmode = VIDMODE_NONE;
	sc->sc_frames_ready = 0;

	if (sc->hw_if->open != NULL) {
		error = sc->hw_if->open(sc->hw_hdl, flags, &sc->sc_fsize,
		    sc->sc_fbuffer, video_intr, sc);
	}
	if (error == 0) {
		sc->sc_open = 1;
		DPRINTF(1, "%s: set device to open\n", __func__);
	}

done:
	device_unref(&sc->dev);
	return (error);
}

int
videoclose(dev_t dev, int flags, int fmt, struct proc *p)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	int error = 0;

	KERNEL_ASSERT_LOCKED();

	DPRINTF(1, "%s: last close\n", __func__);

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (ENXIO);

	error = video_stop(sc);
	sc->sc_open = 0;

	device_unref(&sc->dev);
	return (error);
}

int
videoread(dev_t dev, struct uio *uio, int ioflag)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	int error = 0;
	size_t size;

	KERNEL_ASSERT_LOCKED();

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (ENXIO);

	if (sc->sc_dying) {
		error = EIO;
		goto done;
	}

	if (sc->sc_vidmode == VIDMODE_MMAP) {
		error = EBUSY;
		goto done;
	}

	if ((error = video_claim(sc, curproc->p_p)))
		goto done;

	/* start the stream if not already started */
	if (sc->sc_vidmode == VIDMODE_NONE && sc->hw_if->start_read) {
 		error = sc->hw_if->start_read(sc->hw_hdl);
 		if (error)
			goto done;
		sc->sc_vidmode = VIDMODE_READ;
 	}

	DPRINTF(1, "resid=%zu\n", uio->uio_resid);

	mtx_enter(&sc->sc_mtx);

	if (sc->sc_frames_ready < 1) {
		/* block userland read until a frame is ready */
		error = msleep_nsec(sc, &sc->sc_mtx, PWAIT | PCATCH,
		    "vid_rd", INFSLP);
		if (sc->sc_dying)
			error = EIO;
		if (error) {
			mtx_leave(&sc->sc_mtx);
			goto done;
		}
	}
	sc->sc_frames_ready--;

	mtx_leave(&sc->sc_mtx);

	/* move no more than 1 frame to userland, as per specification */
	size = ulmin(uio->uio_resid, sc->sc_fsize);
	if (!atomic_load_int(&video_record_enable))
		bzero(sc->sc_fbuffer, size);
	error = uiomove(sc->sc_fbuffer, size, uio);
	if (error)
		goto done;

	DPRINTF(1, "uiomove successfully done (%zu bytes)\n", size);

done:
	device_unref(&sc->dev);
	return (error);
}

int
videoioctl(dev_t dev, u_long cmd, caddr_t data, int flags, struct proc *p)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	struct v4l2_buffer *vb = (struct v4l2_buffer *)data;
	int error;

	KERNEL_ASSERT_LOCKED();

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (ENXIO);

	if (sc->hw_if == NULL) {
		error = ENXIO;
		goto done;
	}

	DPRINTF(3, "video_ioctl(%zu, '%c', %zu)\n",
	    IOCPARM_LEN(cmd), (int) IOCGROUP(cmd), cmd & 0xff);

	error = EOPNOTSUPP;
	switch (cmd) {
	case VIDIOC_G_CTRL:
		if (sc->hw_if->g_ctrl)
			error = (sc->hw_if->g_ctrl)(sc->hw_hdl,
			    (struct v4l2_control *)data);
		break;
	case VIDIOC_S_CTRL:
		if (sc->hw_if->s_ctrl)
			error = (sc->hw_if->s_ctrl)(sc->hw_hdl,
			    (struct v4l2_control *)data);
		break;
	default:
		error = (ENOTTY);
	}
	if (error != ENOTTY)
		goto done;

	if ((error = video_claim(sc, p->p_p)))
		goto done;

	/*
	 * The following IOCTLs can only be called by the device owner.
	 * For further shared IOCTLs please move it up.
	 */
	error = EOPNOTSUPP;
	switch (cmd) {
	case VIDIOC_QUERYCAP:
		if (sc->hw_if->querycap)
			error = (sc->hw_if->querycap)(sc->hw_hdl,
			    (struct v4l2_capability *)data);
		break;
	case VIDIOC_ENUM_FMT:
		if (sc->hw_if->enum_fmt)
			error = (sc->hw_if->enum_fmt)(sc->hw_hdl,
			    (struct v4l2_fmtdesc *)data);
		break;
	case VIDIOC_ENUM_FRAMESIZES:
		if (sc->hw_if->enum_fsizes)
			error = (sc->hw_if->enum_fsizes)(sc->hw_hdl,
			    (struct v4l2_frmsizeenum *)data);
		break;
	case VIDIOC_ENUM_FRAMEINTERVALS:
		if (sc->hw_if->enum_fivals)
			error = (sc->hw_if->enum_fivals)(sc->hw_hdl,
			    (struct v4l2_frmivalenum *)data);
		break;
	case VIDIOC_S_FMT:
		if (!(flags & FWRITE))
			return (EACCES);
		if (sc->hw_if->s_fmt)
			error = (sc->hw_if->s_fmt)(sc->hw_hdl,
			    (struct v4l2_format *)data);
		break;
	case VIDIOC_G_FMT:
		if (sc->hw_if->g_fmt)
			error = (sc->hw_if->g_fmt)(sc->hw_hdl,
			    (struct v4l2_format *)data);
		break;
	case VIDIOC_S_PARM:
		if (sc->hw_if->s_parm)
			error = (sc->hw_if->s_parm)(sc->hw_hdl,
			    (struct v4l2_streamparm *)data);
		break;
	case VIDIOC_G_PARM:
		if (sc->hw_if->g_parm)
			error = (sc->hw_if->g_parm)(sc->hw_hdl,
			    (struct v4l2_streamparm *)data);
		break;
	case VIDIOC_ENUMINPUT:
		if (sc->hw_if->enum_input)
			error = (sc->hw_if->enum_input)(sc->hw_hdl,
			    (struct v4l2_input *)data);
		break;
	case VIDIOC_S_INPUT:
		if (sc->hw_if->s_input)
			error = (sc->hw_if->s_input)(sc->hw_hdl,
			    (int)*data);
		break;
	case VIDIOC_G_INPUT:
		if (sc->hw_if->g_input)
			error = (sc->hw_if->g_input)(sc->hw_hdl,
			    (int *)data);
		break;
	case VIDIOC_REQBUFS:
		if (sc->hw_if->reqbufs)
			error = (sc->hw_if->reqbufs)(sc->hw_hdl,
			    (struct v4l2_requestbuffers *)data);
		break;
	case VIDIOC_QUERYBUF:
		if (sc->hw_if->querybuf)
			error = (sc->hw_if->querybuf)(sc->hw_hdl,
			    (struct v4l2_buffer *)data);
		break;
	case VIDIOC_QBUF:
		if (sc->hw_if->qbuf)
			error = (sc->hw_if->qbuf)(sc->hw_hdl,
			    (struct v4l2_buffer *)data);
		break;
	case VIDIOC_DQBUF:
		if (!sc->hw_if->dqbuf)
			break;
		/* should have called mmap() before now */
		if (sc->sc_vidmode != VIDMODE_MMAP) {
			error = EINVAL;
			break;
		}
		error = (sc->hw_if->dqbuf)(sc->hw_hdl,
		    (struct v4l2_buffer *)data);
		if (!atomic_load_int(&video_record_enable))
			bzero(sc->sc_fbuffer_mmap + vb->m.offset, vb->length);
		mtx_enter(&sc->sc_mtx);
		sc->sc_frames_ready--;
		mtx_leave(&sc->sc_mtx);
		break;
	case VIDIOC_STREAMON:
		if (sc->hw_if->streamon)
			error = (sc->hw_if->streamon)(sc->hw_hdl,
			    (int)*data);
		break;
	case VIDIOC_STREAMOFF:
		if (sc->hw_if->streamoff)
			error = (sc->hw_if->streamoff)(sc->hw_hdl,
			    (int)*data);
		if (!error) {
			/* Release device ownership and streaming buffers. */
			error = video_stop(sc);
		}
		break;
	case VIDIOC_TRY_FMT:
		if (sc->hw_if->try_fmt)
			error = (sc->hw_if->try_fmt)(sc->hw_hdl,
			    (struct v4l2_format *)data);
		break;
	case VIDIOC_QUERYCTRL:
		if (sc->hw_if->queryctrl)
			error = (sc->hw_if->queryctrl)(sc->hw_hdl,
			    (struct v4l2_queryctrl *)data);
		break;
	default:
		error = (ENOTTY);
	}

done:
	device_unref(&sc->dev);
	return (error);
}

paddr_t
videommap(dev_t dev, off_t off, int prot)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	caddr_t p;
	paddr_t pa;

	KERNEL_ASSERT_LOCKED();

	DPRINTF(2, "%s: off=%lld, prot=%d\n", __func__, off, prot);

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (-1);

	if (sc->sc_dying)
		goto err;

	if (sc->hw_if->mappage == NULL)
		goto err;

	p = sc->hw_if->mappage(sc->hw_hdl, off, prot);
	if (p == NULL)
		goto err;
	if (pmap_extract(pmap_kernel(), (vaddr_t)p, &pa) == FALSE)
		panic("videommap: invalid page");
	sc->sc_vidmode = VIDMODE_MMAP;

	/* store frame buffer base address for later blanking */
	if (off == 0)
		sc->sc_fbuffer_mmap = p;

	device_unref(&sc->dev);
	return (pa);

err:
	device_unref(&sc->dev);
	return (-1);
}

void
filt_videodetach(struct knote *kn)
{
	struct video_softc *sc = kn->kn_hook;

	klist_remove(&sc->sc_rklist, kn);
}

int
filt_videoread(struct knote *kn, long hint)
{
	struct video_softc *sc = kn->kn_hook;

	if (sc->sc_frames_ready > 0)
		return (1);

	return (0);
}

int
filt_videomodify(struct kevent *kev, struct knote *kn)
{
	struct video_softc *sc = kn->kn_hook;
	int active;

	mtx_enter(&sc->sc_mtx);
	active = knote_modify(kev, kn); 
	mtx_leave(&sc->sc_mtx);

	return (active);
}

int
filt_videoprocess(struct knote *kn, struct kevent *kev)
{
	struct video_softc *sc = kn->kn_hook;
	int active;

	mtx_enter(&sc->sc_mtx);
	active = knote_process(kn, kev);
	mtx_leave(&sc->sc_mtx);

	return (active);
}

const struct filterops video_filtops = {
	.f_flags	= FILTEROP_ISFD | FILTEROP_MPSAFE,
	.f_attach	= NULL,
	.f_detach	= filt_videodetach,
	.f_event	= filt_videoread,
	.f_modify	= filt_videomodify,
	.f_process	= filt_videoprocess,
};

int
videokqfilter(dev_t dev, struct knote *kn)
{
	int unit = VIDEOUNIT(dev);
	struct video_softc *sc;
	int error = 0;

	KERNEL_ASSERT_LOCKED();

	sc = (struct video_softc *)device_lookup(&video_cd, unit);
	if (sc == NULL)
		return (ENXIO);

	if (sc->sc_dying) {
		error = ENXIO;
		goto done;
	}

	switch (kn->kn_filter) {
	case EVFILT_READ:
		kn->kn_fop = &video_filtops;
		kn->kn_hook = sc;
		break;
	default:
		error  = EINVAL;
		goto done;
	}

	if ((error = video_claim(sc, curproc->p_p)))
		goto done;

	/*
	 * Start the stream in read() mode if not already started.  If
	 * the user wanted mmap() mode, he should have called mmap()
	 * before now.
	 */
	if (sc->sc_vidmode == VIDMODE_NONE && sc->hw_if->start_read) {
		if (sc->hw_if->start_read(sc->hw_hdl))
			return (ENXIO);
		sc->sc_vidmode = VIDMODE_READ;
	}

	klist_insert(&sc->sc_rklist, kn);

done:
	device_unref(&sc->dev);
	return (error);
}

int
video_submatch(struct device *parent, void *match, void *aux)
{
        struct cfdata *cf = match;

	return (cf->cf_driver == &video_cd);
}

/*
 * Called from hardware driver. This is where the MI video driver gets
 * probed/attached to the hardware driver
 */
struct device *
video_attach_mi(const struct video_hw_if *rhwp, void *hdlp, struct device *dev)
{
	struct video_attach_args arg;

	arg.hwif = rhwp;
	arg.hdl = hdlp;
	return (config_found_sm(dev, &arg, videoprint, video_submatch));
}

void
video_intr(void *addr)
{
	struct video_softc *sc = (struct video_softc *)addr;

	DPRINTF(3, "video_intr sc=%p\n", sc);
	mtx_enter(&sc->sc_mtx);
	if (sc->sc_vidmode != VIDMODE_NONE)
		sc->sc_frames_ready++;
	else
		printf("%s: interrupt but no streams!\n", __func__);
	if (sc->sc_vidmode == VIDMODE_READ)
		wakeup(sc);
	knote_locked(&sc->sc_rklist, 0);
	mtx_leave(&sc->sc_mtx);
}

int
video_stop(struct video_softc *sc)
{
	int error = 0;

	DPRINTF(1, "%s: stream close\n", __func__);

	if (sc->hw_if->close != NULL)
		error = sc->hw_if->close(sc->hw_hdl);

	sc->sc_vidmode = VIDMODE_NONE;
	mtx_enter(&sc->sc_mtx);
	sc->sc_frames_ready = 0;
	mtx_leave(&sc->sc_mtx);
	sc->sc_owner = NULL;

	return (error);
}

int
video_claim(struct video_softc *sc, struct process *pr)
{
	if (sc->sc_owner != NULL && sc->sc_owner != pr) {
		DPRINTF(1, "%s: already owned=%p\n", __func__, sc->sc_owner);
		return (EBUSY);
	}

	if (sc->sc_owner == NULL) {
		sc->sc_owner = pr;
		DPRINTF(1, "%s: new owner=%p\n", __func__, sc->sc_owner);
	}

	return (0);
}

int
videoprint(void *aux, const char *pnp)
{
	if (pnp != NULL)
		printf("video at %s", pnp);
	return (UNCONF);
}

int
videodetach(struct device *self, int flags)
{
	struct video_softc *sc = (struct video_softc *)self;
	int maj, mn;

	/* locate the major number */
	for (maj = 0; maj < nchrdev; maj++)
		if (cdevsw[maj].d_open == videoopen)
			break;

	/* Nuke the vnodes for any open instances (calls close). */
	mn = self->dv_unit;
	vdevgone(maj, mn, mn, VCHR);

	klist_invalidate(&sc->sc_rklist);
	klist_free(&sc->sc_rklist);

	free(sc->sc_fbuffer, M_DEVBUF, sc->sc_fbufferlen);

	return (0);
}

int
videoactivate(struct device *self, int act)
{
	struct video_softc *sc = (struct video_softc *)self;

	switch (act) {
	case DVACT_DEACTIVATE:
		sc->sc_dying = 1;
		break;
	}
	return (0);
}
